### PR TITLE
Implement a multi-app approach to JSX-loading

### PR DESCRIPTION
**Pull request by mBeierl**
This structure is just a suggestion of mine that would allow JSX code to be generalized for all Adobe Apps and only put App-specific code in the files of the appropriate folders (e.g. jsx/PPRO for Premiere Pro).

I've been working with the PProPanel a lot, that's I'm applying it to this sample, although this would be a better fit for something like a "best practices" sample panel.

Generally, would you agree that this is a good approach for cross-app panel development?

### CODE REVIEW

**Comment by bbb999** [comment]
Looks fine to me!

File: PProPanel/ext.js [hunk 1]
```diff
@@ -222,11 +222,11 @@ function loadJSX() {
 	var extensionPath = csInterface.getSystemPath(SystemPath.EXTENSION);
 
 	// load general JSX script independent of appName
-	const extensionRootGeneral = extensionPath + '/jsx/';
+	var extensionRootGeneral = extensionPath + '/jsx/';
```
Comment (ericdrobinson) [inline PProPanel/ext.js]: Out of curiosity, why was this changed from `const` to `var`?

File: PProPanel/jsx/general.jsx [hunk 2]
```diff
@@ -0,0 +1,6 @@
+if (typeof $ == 'undefined') {
```
Comment (ericdrobinson) [inline PProPanel/jsx/general.jsx]: Is this actually ever the case for ExtendScript contexts? I was under the impression that $ was a built-in global (see: Object Model Viewer or page 216 of the [JavaScript Tools Guide](http://wwwimages.adobe.com/content/dam/Adobe/en/devnet/scripting/pdfs/javascript_tools_guide.pdf)).

Comment (bbb999) [inline PProPanel/jsx/general.jsx]: [That is correct.]

Comment (ericdrobinson) [inline PProPanel/jsx/general.jsx]: Why this check, then? A comment _at the very least_ as to its purpose is desirable.

I understand that, generally, `$` is jQuery in JavaScript. My _guess_ is that this is a hyper-conservative way to ensure that no one "renamed" the ExtendScript `$` Helper object and didn't somehow replace it with something else. _If_ this is the case, then some comments about this fact would be very helpful for those trying to piece together how the environment actually works.

The same holds true for [PProPanel.jsx](https://github.com/mBeierl/Samples/blob/d53d277874b951eb57d46ffee6a658748663ebf5/PProPanel/PProPanel.jsx), actually. `$` is a super important object in ExtendScript contexts and to the uninitiated, that code simply looks like you might be grafting something onto jQuery or another library...

Comment (mBeierl) [inline PProPanel/jsx/general.jsx]: It was part of samples before, so I kept it. In general it is always better to have it checked than to get an error. Especially in Extendscript :-P 

@bbb999 is the $ object always injected? That you put "that is correct" in [ ] let's me assume that you're actually not 100% sure ;-)

File: PProPanel/ext.js (hunk 1 above)
Comment (mBeierl) [inline PProPanel/ext.js]: Because const is an [ES6 standard](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Statements/const), and I don't think it's supported by the CEF version used in CEP. 

But I could be mistaken, haven't tried it yet. In my projects I use babel to transpile my source code, that's why I used const there in the first place. 
This sample doesn't use any build setup, so I thought I'd better keep it clean.

File: PProPanel/jsx/general.jsx (hunk 2 above)
Comment (ericdrobinson) [inline PProPanel/jsx/general.jsx]: >@bbb999 is the $ object always injected?
While I am not @bbb999, I will say that `$` is a global object defined in the ExtendScript extensions to JavaScript. It is part of the environment. It _may_ be possible to **reassign** it without replacing it, but I've never seen an example of this in ExtendScript.

I would suggest that these files get changed such that the check that `$` exists be replaced with a comment along the lines of:

```javascript
// Extensions defined herein are grafted onto ExtendScript's global "$" object.
```

File: PProPanel/ext.js (hunk 1 above)
Comment (ericdrobinson) [inline PProPanel/ext.js]: >Because const is an [ES6 standard](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Statements/const), and I don't think it's supported by the CEF version used in CEP.

Ahh. This probably depends a bit on what the minimum supported CEP version should be for the PProPanel sample. @bbb999: how far back in the release history is the sample intended to support?

@mBeierl You can easily find this out in [the docs](https://github.com/Adobe-CEP/CEP-Resources), but CEP 5.2 is equivalent to **Chrome 27**, while CEP 6.1+ is equivalent to **Chrome 41**. [This](http://caniuse.com/#compare=chrome+27,chrome+41) is a quick view comparing what does and doesn't work in those browsers. If you check [`const` support specifically](http://caniuse.com/#feat=const) you can see that there are some extra notes:

* **Chrome 27** - `const` is recognized only when _not_ in `strict` mode. No block scoping.
* **Chrome 41** - `const` has block scoping in `strict` mode, but no block scoping otherwise.

As `strict` mode is _not_ used in the scripts in question, `const` would be supported identically - it works, just not in a block-scoped fashion. **_Personally_**, I would suggest making use of it where possible as it is a good enhancement and improves readability. I'd ask @bbb999 to weigh in on this, though.

>In my projects I use babel to transpile my source code, that's why I used const there in the first place.

Have you checked out the [TypeScript sample](https://github.com/Adobe-CEP/Samples/tree/master/TypeScript) yet? The environment there is specifically set up to provide you with autocompletion and IntelliSense support for ExtendScript files (via [type declaration files](https://github.com/Adobe-CEP/Samples/tree/master/TypeScript/PProPanel-vscode/dom_app/types). It's also pre-configured (see the [`dom_html/tsconfig.json` file](https://github.com/Adobe-CEP/Samples/blob/master/TypeScript/PProPanel-vscode/dom_html/tsconfig.json)) to transpile _both_ TypeScript and JavaScript (`--allowJs` into panel-compatible JavaScript (es5 target, with a note to switch to es3 if issues arise).

Comment (mBeierl) [inline PProPanel/ext.js]: > Have you checked out the TypeScript sample yet?

No, I didn't check it out yet. Personally, I don't like to work with TypeScript, but I will have a look at the sample in this repo.

For my panel projects, I've set up a boilerplate using [create-react-app](https://github.com/facebookincubator/create-react-app), because reactJS is the ui framework we use at work.

> Chrome 27 - const is recognized only when not in strict mode. No block scoping.
> Chrome 41 - const has block scoping in strict mode, but no block scoping otherwise.

Hmm... the fact that ES specs differ between versions of CEP really just drives me to hold on to the extra compile step using babel, as this will maximize compatibility and it allows me to use the new cool things of JavaScript like arrow functions, spread operator and so on ;-)

File: PProPanel/jsx/general.jsx (hunk 2 above)
Comment (mBeierl) [inline PProPanel/jsx/general.jsx]: Well it doesn't hurt to do the extra check, if there's already a $ object defined than it won't replace it.

On the other hand we could also just get rid of it if it's really not needed. I just included it here because it was done by @bbb999 in the [PProPanel.jsx](https://github.com/Adobe-CEP/Samples/blob/master/PProPanel/PProPanel.jsx) file.

File: PProPanel/ext.js (hunk 1 above)
Comment (ericdrobinson) [inline PProPanel/ext.js]: >For my panel projects, I've set up a boilerplate using create-react-app, because reactJS is the ui framework we use at work.

Neat. We use [Vue](https://vuejs.org/).

>Hmm... the fact that ES specs differ between versions of CEP really just drives me to hold on to the extra compile step using babel, as this will maximize compatibility and it allows me to use the new cool things of JavaScript like arrow functions, spread operator and so on ;-)

I hear you there, but you're not committing to an environment that runs with babel. The environment you're committing to _does_ support `const` (especially as used within this context - i.e. non-strict mode), so it would be preferable to make use of it. :) (Pending, of course, direction from @bbb999...)

File: PProPanel/jsx/general.jsx (hunk 2 above)
Comment (ericdrobinson) [inline PProPanel/jsx/general.jsx]: >Well it doesn't hurt to do the extra check, if there's already a $ object defined then it won't replace it.

I agree with the exception that it _suggests_ that the `$` object is similar to that in pure JavaScript - something libraries tend to use. In ExtendScript it is something special and always included by default unless you did something strange. It raises questions.

Likewise, though, I would get @bbb999's thoughts before making such a change :D

File: PProPanel/ext.js (hunk 1 above)
Comment (mBeierl) [inline PProPanel/ext.js]: Alright, it would certainly refresh this project to a more modern JS standard ;-)

My reason for not including it in this PR: I wanted to focus on the JSX structure change and keep the code style to what is the current standard.

I'd suggest a new PR that contains the switch to const (& maybe 'let'?) where it makes sense. It should update all the JS code to a more modern JS standard that is compliant with the CEP environment.

Comment (ericdrobinson) [inline PProPanel/ext.js]: @mBeierl That would be _awesome_. [`let` is a bit of an odd case](http://caniuse.com/#feat=let) as it would require enabling experimental JavaScript features on Chrome 27 (CEP 5.2-6.0) and `strict` mode for Chrome 41...

On a side note, I just [posted](https://forums.adobe.com/thread/2353067) a bit of an explanation about the TypeScript Sample to Adobe's forums. Perhaps that would be a good place to start to understand the "Why" behind the setup.
